File: src/ProjetoCanil/Modelo/Adotavel.cs
namespace ProjetoCanil.Modelo;

/// <summary>
/// Define o contrato para animais que podem ser adotados no sistema.
/// </summary>
public interface IAdotavel
{
    /// <summary>
    /// Obtém o nome do animal.
    /// </summary>
    string Nome { get; }

    /// <summary>
    /// Verifica se o animal está disponível para adoção.
    /// </summary>
    /// <returns><c>true</c> se disponível, <c>false</c> caso contrário.</returns>
    bool EstaDisponivel();

    /// <summary>
    /// Atualiza o status de disponibilidade do animal.
    /// </summary>
    /// <param name="disponivel">Novo status.</param>
    void SetDisponivel(bool disponivel);

    /// <summary>
    /// Executa o processo de adoção, alterando o status para indisponível.
    /// </summary>
    void ProcessoAdocao()
    {
        SetDisponivel(false);
        Console.WriteLine($"{Nome} FOI ADOTADO!");
    }
}

File: legacy-java/modelo/Adotavel.java
package Bpoo.projeto_canil.modelo;

/**
 * Define o contrato para animais que podem ser adotados no sistema.
 */
public interface Adotavel {

    /**
     * Obtém o nome do animal.
     * @return nome do animal.
     */
    String getNome();

    /**
     * Verifica se o animal está disponível para adoção.
     * @return {@code true} se disponível, {@code false} caso contrário.
     */
    boolean estaDisponivel();

    /**
     * Atualiza o status de disponibilidade do animal.
     * @param disponivel novo status.
     */
    void setDisponivel(boolean disponivel);

    /**
     * Executa o processo de adoção, alterando o status para indisponível.
     */
    default void processoAdocao() {
        setDisponivel(false);
        System.out.println(getNome() + " FOI ADOTADO!");
    }
}
